recipes-openxt/xenclient-tpm-scripts/xenclient-tpm-scripts/tpm-functions

```diff
@@ -619,26 +619,128 @@ tpm2_write_tboot_policy() {
                  "${policy}" >/dev/null 2>&1
 }
 
+# Usage: tpm_get_syspath [TPM-DEVICE-ID]
+# Verify and print the path of the TPM, with the given id (default:0), on
+# stdout. Returns 0 on success, 1 if the path in the sysfs cannot be found.
+tpm_get_syspath() {
+    local devid="${1:-0}"
+    local syspath="/sys/class/tpm/tpm${devid}"
+
+    if [ ! -e "${syspath}" ]; then
+        echo "tpm_get_syspath: tpm${devid} has no entry in the sysfs." >&2
+        return 1
+    fi
+    echo "${syspath}"
+    return 0
+}
+
+# Usage: tpm_get_ev_separator
+# Print the EV_SEPARATOR value as presented by the kernel, or our best guess as
+# what it may be, on stdout.
+tpm_get_ev_separator() {
+    local evs
+
+    if is_tpm_2_0; then
+        # TPM 2.0: `printf "\xff\xff\xff\xff" | sha256sum`, best guess...
+        evs="df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119"
+     else
+        # TPM 1.2 will expose a securityfs node with ASCII values.
+        local path="/sys/kernel/security/tpm0/ascii_bios_measurements"
+
+        evs=$(awk '$1 == 4 && $3 == 04 { print $2 }' "${path}")
+        # Fallback: Pick `printf "\xff\xff\xff\xff" | sha1sum`, best guess...
+        evs=${evs:-"d9be6524a5f5047db5866813acf3277892a7a30a"}
+    fi
+    echo "${evs}"
+}
+
+# Usage: tpm_list_pcrs
+# Print all the PCR values of the first TPM on stdout.
+tpm_list_pcrs() {
+    if is_tpm_2_0; then
+        tpm2_pcrlist
+    else
+        cat "$(tpm_get_syspath)/device/pcrs"
+    fi
+}
+
+# tpm_pcrlist_save PCRLIST-FILE PCR-IDS
+# Write the list of PCR-IDS to the PCRLIST-FILE, overwriting the existing one
+# if necessary.
+tpm_pcrlist_save() {
+    local file="$1"
+    shift 1
+    local pcr
+    local pcrlist=""
+
+    # Sanity check list of number in PCR id range ([0-23])
+    for p in $@; do
```

Comment:
Usually you want "$@" to ensure parameters aren't space split again. `foo "1 2" "3 4"` has `1="1 2"` and `2="3 4"`.  It would be wrong to call this function that way, but it stood out to me.

Reply:
Absolutely, these quotes are missing.